core/modules/util/EventThread.cc

```diff
@@ -0,0 +1,164 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2015 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ *  @author: John Gates,
+ */
+
+// Class header
+#include "util/EventThread.h"
+
+// System headers
+#include <algorithm>
+
+// Third-party headers
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+namespace lsst {
+namespace qserv {
+namespace util {
+
+/// Handle commands as they arrive until queEnd() is called.
+void EventThread::handleCmds() {
+    startup();
+    while(_loop) {
+        Command::Ptr cmd = _q->getCmd();
+        if (cmd != nullptr) {
+            _q->commandStart(cmd);
+            if (cmd->action() == HALT) {
+                _loop = false;
+            }
+            _q->commandFinish(cmd);
+            cmd->resetFunc(); // Reset _func in case it has a captured Command::Ptr
+        }
+    }
+    finishup();
+}
+
+/// call this to start the thread
+void EventThread::run() {
+    std::thread t{&EventThread::handleCmds, this};
+    _t = move(t);
+}
+
+PoolEventThread::~PoolEventThread() {
+    LOGF_DEBUG("PoolEventThread::~PoolEventThread()");
+}
+
+void PoolEventThread::finishup() {
+    //  Need to make sure the thread data doesn't disappear before the thread finishes.
+    PoolEventThread::Ptr pet = _threadPool->release(this);
+    if (pet != nullptr) {
+        auto f = [pet](){ pet->join(); };
+        std::thread t{f};
+        t.detach();
```

Comment:
I think I'm probably being overly cautious here.  'pet' is probably the only thing keeping the PoolEventThread object alive after _threadPool->release is called (but it may not be). I don't really expect anything to be called after finishup(), but I'd rather waste a few cpu cycles than segfault as the captured 'pet' will keep the object alive until it joins. I don't expect this code to be called often. I can't move _threadPool->release to the destructor as it will never get called. It's wonky but safe.